### PR TITLE
Added Password Validators

### PR DESCRIPTION
**Pull request by imhassantariq**
#### Story Link
[Review Password Requirements](https://edlyio.atlassian.net/browse/EDE-384)

#### PR Description

Added two of the password validators:
- MasterPasswordValidator
- RestrictedSymbolValidator

#### Type of change

Please select the options that are relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation Change

#### How to test?

For testing this ticket you need to set the value of your `AUTH_PASSWORD_VALIDATORS` (located in `lms.env.json`) like this:

```
"AUTH_PASSWORD_VALIDATORS":[
    {
        "NAME":"util.password_policy_validators.MaximumLengthValidator",
        "OPTIONS":{
            "max_length":75
        }
    },
    {
        "NAME":"util.password_policy_validators.MinimumLengthValidator",
        "OPTIONS":{
            "min_length":8
        }
    },
    {
        "NAME":"openedx.features.ucsd_features.validators.RestrictedSymbolValidator",
        "OPTIONS":{
            "restricted_symbol_list":["$", "'", "!", ":", " "]
        }
    },
    {
        "NAME":"openedx.features.ucsd_features.validators.MasterPasswordValidator",
        "OPTIONS":{
            "min_validations":3,
            "validators":[
                {
                    "NAME":"util.password_policy_validators.LowercaseValidator",
                    "OPTIONS":{
                        "min_lower":1
                    }
                },
                {
                    "NAME":"util.password_policy_validators.UppercaseValidator",
                    "OPTIONS":{
                        "min_upper":1
                    }
                },
                {
                    "NAME":"util.password_policy_validators.NumericValidator",
                    "OPTIONS":{
                        "min_numeric":1
                    }
                },
                {
                    "NAME":"django.contrib.auth.password_validation.UserAttributeSimilarityValidator"
                },
                {
                    "NAME":"util.password_policy_validators.PunctuationValidator",
                    "OPTIONS":{
                        "min_punctuation":1
                    }
                }
            ]
        }
    }
]
```

#### Checklist before merging:

- [ ] Squased
- [ ] Reviewd

### CODE REVIEW

**Comment by edtecheco** [comment]
Your PR has finished running tests. There were no failures.

File: openedx/features/ucsd_features/validators.py [hunk 1]
```diff
@@ -0,0 +1,125 @@
...
+                passed += 1
+        if passed < self.min_validations:
+            errors.insert(0, ValidationError(" Fix at least {} of the following errors: ".format(
+                self.min_validations - passed)))
+            raise ValidationError(errors, code="too_few_validations",
+                                  params={"min_validations": self.min_validations})
+
+    def get_help_text(self):
+        text = ungettext(
+            "your password must contain at least %(min_validations)d of the following:",
```
Comment (danialmalik) [inline openedx/features/ucsd_features/validators.py:52]:
```suggestion
            "your password must conform to at least %(min_validations)d of the following:",
```

File: openedx/features/ucsd_features/validators.py [hunk 2]
```diff
@@ -0,0 +1,125 @@
...
+            if hasattr(validator, 'get_instruction_text'):
+                text += validator.get_instruction_text() + ","
+        return mark_safe(text[:-1])
+
+    @staticmethod
+    def check_validators_list(validators, min_validations):
+        if len(validators) >= min_validations >= 0:
+            return
+        elif len(validators) < min_validations:
+            raise Exception('Number of Validators in list is lower the the minimum number of required validations.')
```
Comment (danialmalik) [inline openedx/features/ucsd_features/validators.py:76]:
```suggestion
        if len(validators) < min_validations:
            raise Exception('Number of Validators in list is lower the the minimum number of required validations.')
```
From what I understand from the code, we only need a single if block.

File: openedx/features/ucsd_features/validators.py [hunk 3]
```diff
@@ -0,0 +1,125 @@
...
+        if self.validate_restriction_list(password):
+            return
+        raise ValidationError(
+            _("Your password must not contain any of the following symbol: {}".format(self.restricted_symbol_text)),
+            code="restricted_symbol_used",
+        )
+
+    def get_help_text(self):
+        return _(
+            "Your password must not contain any of the following symbol: {}".format(self.restricted_symbol_text))
```
Comment (danialmalik) [inline openedx/features/ucsd_features/validators.py:100]:
```suggestion
            "Your password must not contain any of the following symbols: {}".format(self.restricted_symbol_text))
```

File: openedx/features/ucsd_features/validators.py [hunk 4]
```diff
@@ -0,0 +1,125 @@
...
+
+    def get_instruction_text(self):
+        if len(self.restricted_symbol_list) > 0:
+            return _(
+                "your password must not contain any of the following symbol: {}".format(self.restricted_symbol_text))
+        else:
+            return ""
```
Comment (danialmalik) [inline openedx/features/ucsd_features/validators.py:107]:
```suggestion
        return ""
```

File: openedx/features/ucsd_features/validators.py (hunk 1 above)
Comment (imhassantariq) [reply openedx/features/ucsd_features/validators.py]: Done

File: openedx/features/ucsd_features/validators.py (hunk 2 above)
Comment (imhassantariq) [reply openedx/features/ucsd_features/validators.py]: Done

File: openedx/features/ucsd_features/validators.py (hunk 3 above)
Comment (imhassantariq) [reply openedx/features/ucsd_features/validators.py]: Done

File: openedx/features/ucsd_features/validators.py (hunk 4 above)
Comment (imhassantariq) [reply openedx/features/ucsd_features/validators.py:105]: Done

File: openedx/features/ucsd_features/validators.py [hunk 5]
```diff
@@ -0,0 +1,122 @@
...
+
+    def get_instruction_text(self):
+        text = ungettext(
+            "your password must confirm to at least %(min_validations)d of the following:",
+            "your password must confirm to at least %(min_validations)d of the followings:",
+            self.min_validations
+        ) % {"min_validations": self.min_validations}
+        for validator in self.validators:
+            if hasattr(validator, 'get_instruction_text'):
+                text += validator.get_instruction_text() + ","
```
Comment (HamzaIbnFarooq) [inline openedx/features/ucsd_features/validators.py:68]: nit:
```suggestion
                text += validator.get_instruction_text() + ", "
```

File: openedx/features/ucsd_features/validators.py [hunk 6]
```diff
@@ -0,0 +1,122 @@
...
+            "your password must confirm to at least %(min_validations)d of the followings:",
+            self.min_validations
+        ) % {"min_validations": self.min_validations}
+        for v in self.validators:
+            text += v.get_help_text()
+        return mark_safe(text)
+
+    def get_instruction_text(self):
+        text = ungettext(
+            "your password must confirm to at least %(min_validations)d of the following:",
```
Comment (tehreem-sadat) [inline openedx/features/ucsd_features/validators.py:62]: Repeated lines (62 and 63).

File: openedx/features/ucsd_features/validators.py [hunk 7]
```diff
@@ -0,0 +1,122 @@
...
+    def get_restricted_symbol_text(self):
+        if len(self.restricted_symbol_list) > 0:
+            text = ""
+            for c in self.restricted_symbol_list:
+                if c.isspace():
+                    text += " Space"
+                else:
+                    text += " " + c
+            return text
+        return ""
```
Comment (HamzaIbnFarooq) [inline openedx/features/ucsd_features/validators.py:122]:
```suggestion
        text = ""
        for c in self.restricted_symbol_list:
            if c.isspace():
                text += " Space"
            else:
                text += " " + c
        return text
```

File: openedx/features/ucsd_features/validators.py (hunk 6 above)
Comment (imhassantariq) [reply openedx/features/ucsd_features/validators.py:62]: Singular and Plural at the end `following` and `followings`. Reference: https://github.com/ucsd-ets/edx-platform/blob/develop/common/djangoapps/util/password_policy_validators.py#L177

File: openedx/features/ucsd_features/validators.py (hunk 7 above)
Comment (imhassantariq) [reply openedx/features/ucsd_features/validators.py]: Done

File: openedx/features/ucsd_features/validators.py (hunk 5 above)
Comment (imhassantariq) [reply openedx/features/ucsd_features/validators.py]: Done

File: openedx/features/ucsd_features/validators.py [hunk 8]
```diff
@@ -0,0 +1,120 @@
...
+    Parameters:
+        validators (list): the list of validators to validate along with their options.
+        min_validations (int): the num of minimum validation must pass for password.
+
+    Raises:
+        ValidationError if minimum number of validations is not passed.
+    """
+    def __init__(self, validators=[], min_validations=0):
+        self.check_validators_list(validators, min_validations)
+        self.validators_list = validators
```
Comment (danialmalik) [inline openedx/features/ucsd_features/validators.py:24]: If we are not going to use this list anywhere, I don't think we need to save this list as `self` attribute?

File: openedx/features/ucsd_features/validators.py [hunk 9]
```diff
@@ -0,0 +1,120 @@
...
+        self.validators = []
+        for validator in self.validators_list:
+            try:
+                validator_name = validator["NAME"]
+            except KeyError:
+                raise KeyError("Validator doesn't contains 'NAME'. Please enter valid Validator")
```
Comment (danialmalik) [inline openedx/features/ucsd_features/validators.py:30]:
```suggestion
                raise KeyError("Validator doesn't contain 'NAME'. Please enter valid Validator")
```

File: openedx/features/ucsd_features/validators.py (hunk 4 above)
Comment (danialmalik) [reply openedx/features/ucsd_features/validators.py:105]: Don't think so 😅

File: openedx/features/ucsd_features/validators.py (hunk 8 above)
Comment (imhassantariq) [reply openedx/features/ucsd_features/validators.py:24]: Done

File: openedx/features/ucsd_features/validators.py (hunk 9 above)
Comment (imhassantariq) [reply openedx/features/ucsd_features/validators.py]: Done

File: openedx/features/ucsd_features/validators.py (hunk 4 above)
Comment (imhassantariq) [reply openedx/features/ucsd_features/validators.py]: Done

**Comment by edtecheco** [comment]
Your PR has finished running tests. The following contexts failed:
* jenkins/python

**Comment by imhassantariq** [comment]
jenkins run all

**Comment by edtecheco** [comment]
Your PR has finished running tests. The following contexts failed:
* jenkins/python

**Comment by imhassantariq** [comment]
jenkins run all

**Comment by edtecheco** [comment]
Your PR has finished running tests. The following contexts failed:
* jenkins/python

**Comment by imhassantariq** [comment]
jenkins run all

**Comment by edtecheco** [comment]
Your PR has finished running tests. The following contexts failed:
* jenkins/python

**Pull request merged**